src/backend/distributed/utils/colocation_utils.c

```diff
@@ -14,14 +14,322 @@
+#include "distributed/worker_protocol.h"
 #include "utils/fmgroids.h"
+#include "utils/lsyscache.h"
 #include "utils/rel.h"
 
 
+/* local function forward declarations */
+static void MarkTablesColocated(Oid sourceRelationId, Oid targetRelationId);
+static void ErrorIfShardPlacementsNotColocated(Oid leftRelationId, Oid rightRelationId);
+static bool ShardsIntervalsEqual(ShardInterval *leftShardInterval,
+								 ShardInterval *rightShardInterval);
+static void UpdateColocationGroup(Oid distributedRelationId, uint32 colocationId);
+
+
+/* exports for SQL callable functions */
+PG_FUNCTION_INFO_V1(mark_tables_colocated);
+
+
+/*
+ * mark_tables_colocated puts target tables to same colocation group with the
+ * source table. If the source table is in INVALID_COLOCATION_ID group, then it
+ * creates a new colocation group and assigns all tables to this new colocation
+ * group.
+ */
+Datum
+mark_tables_colocated(PG_FUNCTION_ARGS)
+{
+	Oid sourceRelationId = PG_GETARG_OID(0);
+	ArrayType *relationIdArrayObject = PG_GETARG_ARRAYTYPE_P(1);
+	Datum *relationIdDatumArray = NULL;
+	int index = 0;
+
+	int relationCount = ArrayObjectCount(relationIdArrayObject);
+	if (relationCount < 1)
+	{
+		ereport(ERROR, (errmsg("at least one target table is required for this "
+							   "operation")));
+	}
+
+	relationIdDatumArray = DeconstructArrayObject(relationIdArrayObject);
+
+	for (index = 0; index < relationCount; index++)
+	{
+		Oid nextRelationOid = DatumGetObjectId(relationIdDatumArray[index]);
+		MarkTablesColocated(sourceRelationId, nextRelationOid);
+	}
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * MarkTablesColocated puts both tables to same colocation group. If the
+ * source table is in INVALID_COLOCATION_ID group, then it creates a new
+ * colocation group and assigns both tables to same colocation group. Otherwise,
+ * it adds the target table to colocation group of the source table.
+ */
+static void
+MarkTablesColocated(Oid sourceRelationId, Oid targetRelationId)
+{
+	uint32 sourceColocationId = INVALID_COLOCATION_ID;
+	Relation pgDistColocation = NULL;
+	Var *sourceDistributionColumn = NULL;
+	Var *targetDistributionColumn = NULL;
+	Oid sourceDistributionColumnType = InvalidOid;
+	Oid targetDistributionColumnType = InvalidOid;
+
+	CheckDistributedTable(sourceRelationId);
+	CheckDistributedTable(targetRelationId);
+
+	CheckHashPartitionedTable(sourceRelationId);
+	CheckHashPartitionedTable(targetRelationId);
+
+	sourceDistributionColumn = PartitionKey(sourceRelationId);
+	sourceDistributionColumnType = sourceDistributionColumn->vartype;
+
+	targetDistributionColumn = PartitionKey(targetRelationId);
+	targetDistributionColumnType = targetDistributionColumn->vartype;
+
+	if (sourceDistributionColumnType != targetDistributionColumnType)
+	{
+		char *sourceRelationName = get_rel_name(sourceRelationId);
+		char *targetRelationName = get_rel_name(targetRelationId);
+
+		ereport(ERROR, (errmsg("distribution column types don't match for "
```

Comment:
I think this is more of an error detail. Maybe replace the error message to `cannot colocate tables %s %s` and use this message in the detail.